In [1]:
# Initialize OK
from client.api.notebook import Notebook
ok = Notebook('lab02.ok')

Assignment: lab02
OK, version v1.13.11



## Lab 2: Pandas Continued

**This lab is due Thursday, June 27 at 11:59 PM.**

We will continue discussion of [Pandas](https://pandas.pydata.org/). In this lab, you will learn about:

* Grouping dataframes
* Merging dataframes

In [2]:
# Setup from Discussion 2

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import requests
import zipfile
from pathlib import Path
%matplotlib inline


def fetch_and_cache(data_url, file, data_dir="data", force=False):
    """
    Download and cache a url and return the file object.
    
    data_url: the web address to download
    file: the file in which to save the results.
    data_dir: (default="data") the location to save the data
    force: if true the file is always re-downloaded 
    
    return: The pathlib.Path to the file.
    """
    data_dir = Path(data_dir)
    data_dir.mkdir(exist_ok=True)
    file_path = data_dir/Path(file)
    if force and file_path.exists():
        file_path.unlink()
    if force or not file_path.exists():
        print('Downloading...', end=' ')
        resp = requests.get(data_url)
        with file_path.open('wb') as f:
            f.write(resp.content)
        print('Done!')
    else:
        import time 
        created = time.ctime(file_path.stat().st_ctime)
        print("Using cached version downloaded at", created)
    return file_path

data_url = 'https://www.ssa.gov/oact/babynames/state/namesbystate.zip'
namesbystate_path = fetch_and_cache(data_url, 'namesbystate.zip')
zf = zipfile.ZipFile(namesbystate_path, 'r')

column_labels = ['State', 'Sex', 'Year', 'Name', 'Count']

def load_dataframe_from_zip(zf, f):
    with zf.open(f) as fh: 
        return pd.read_csv(fh, header=None, names=column_labels)

states = [
    load_dataframe_from_zip(zf, f)
    for f in sorted(zf.filelist, key=lambda x:x.filename) 
    if f.filename.endswith('.TXT')
]

baby_names = states[0]
for state_df in states[1:]:
    baby_names = pd.concat([baby_names, state_df])
baby_names = baby_names.reset_index().iloc[:, 1:]
ca = baby_names[baby_names['State'] == 'CA']

Using cached version downloaded at Thu Jun 27 11:10:00 2019



## Data Aggregration (Grouping Data Frames)

### Question 1a
To count the number of instances of each unique value in a `Series`, we can use the `value_counts()` [method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html) as `df["col_name"].value_counts()`. Count the number of different names for each Year in `CA` (California).  (You may use the `ca` DataFrame created above.)

**Note:** *We are not computing the number of babies but instead the number of names (rows in the table) for each year.*

<!--
BEGIN QUESTION
name: q1a
-->

In [5]:
num_of_names_per_year = ca['Year'].value_counts()
#num_of_names_per_year
num_of_names_per_year[:5]
#ca

2007    7250
2008    7158
2009    7119
2006    7075
2010    7010
Name: Year, dtype: int64

In [6]:
ok.grade("q1a");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 4
    Failed: 0
[ooooooooook] 100.0% passed



### Question 1b
Count the number of different names for each gender in `CA`.


<!--
BEGIN QUESTION
name: q1b
-->

In [7]:
num_of_names_per_gender = ca['Sex'].value_counts()
num_of_names_per_gender
#ca

F    224810
M    156404
Name: Sex, dtype: int64

In [8]:
ok.grade("q1b");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



### Question 2: Groupby ###

Before we jump into using the [`groupby`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html) function in Pandas, let's recap how grouping works in general for tabular data through a guided set of questions based on a small toy dataset of movies and genres. 

**Note:** If you want to see a visual of how grouping of data works, here is a link to an animation [Groupby Animation](http://www.ds100.org/sp18/assets/lectures/lec03/03-groupby_and_pivot.pdf)

**Problem Setting:** This summer 2018, there were a lot of good and bad movies that came out. Below is a dataframe with 5 columns: name of the movie as a `string`, the genre of the movie as a `string`, the first name of the director of the movie as a `string`, the average rating out of 10 on Rotten Tomatoes as an `integer`, and the total gross revenue made by the movie as an `integer`. The point of these guided questions (parts a and b) below is to understand how grouping of data works in general, **not** how grouping works in code. We will worry about how grouping works in Pandas in 7c, which will follow.

Below is the `movies` dataframe we are using, imported from the `movies.csv` file located in the `lab02` directory.

In [9]:
movies = pd.read_csv("movies.csv")
movies

director                      genre  \
0         David         Action & Adventure   
1          Bill                     Comedy   
2           Ron  Science Fiction & Fantasy   
3      Baltasar                      Drama   
4          Bart                      Drama   
5          Gary         Action & Adventure   
6          Drew         Action & Adventure   
7          Brad                  Animation   
8          Jeff                     Comedy   
9          J.A.  Science Fiction & Fantasy   
10      Charles                     Comedy   
11       Gerard                     Horror   
12       Peyton         Action & Adventure   
13       Genndy                  Animation   
14       Rawson         Action & Adventure   
15           Ol                     Comedy   
16  Christopher         Action & Adventure   
17         Marc                     Comedy   

                                    movie  rating    revenue  
0                              Deadpool 2       7  318344544  
1                               Book Club       5   68566296  
2                 Solo: A Star Wars Story       6  213476293  
3                                  Adrift       6   31445012  
4                        American Animals       6    2847319  
5                                Oceans 8       6  138803463  
6                           Hotel Artemis       8    6708147  
7                           Incredibles 2       5  594398019  
8                                     Tag       6   54336863  
9          Jurassic World: Fallen Kingdom       6  411873505  
10                             Uncle Drew       5   42201656  
11                        The First Purge       7   68765655  
12                   Ant-Man and the Wasp       5  208681866  
13  Hotel Transylvania 3: Summer Vacation       5  154418311  
14                             Skyscraper       6   66801215  
15            Mamma Mia! Here We Go Again       8  111705055  
16            Mission: Impossible-Fallout       6  182080372  
17                     Christopher Robbin       6    6786317

### Question 2a

If we grouped the `movies` dataframe above by `genre`, how many groups would be in the output and what would be the groups? Assign `num_groups` to the number of groups created (hard-code) and fill in `genre_list` as a list containing the names of genres as strings that represent the groups.

<!--
BEGIN QUESTION
name: q2a
-->

In [10]:
num_groups = len((movies['genre'].value_counts()))
#num_groups
genre_list = movies.groupby(movies['genre']).count().reset_index()['genre'].tolist()


genre_list

['Action & Adventure',
 'Animation',
 'Comedy',
 'Drama',
 'Horror',
 'Science Fiction & Fantasy']

In [11]:
ok.grade("q2a");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



### Question 2b

Whenever we group tabular data, it is usually the case that we need to aggregate values from the ungrouped column(s). If we were to group the `movies` dataframe above by `genre`, which column(s) in the `movies` dataframe would it make sense to aggregate if we were interested in finding how well each genre did in the eyes of people? Fill in `agg_cols` with the column name(s).

<!--
BEGIN QUESTION
name: q2b
-->

In [80]:
agg_cols = ['rating','revenue']

State Sex  Year       Name  Count
379228    CA   F  1910       Mary    295
379229    CA   F  1910      Helen    239
379230    CA   F  1910    Dorothy    220
379231    CA   F  1910   Margaret    163
379232    CA   F  1910    Frances    134
379233    CA   F  1910       Ruth    128
379234    CA   F  1910     Evelyn    126
379235    CA   F  1910      Alice    118
379236    CA   F  1910   Virginia    101
379237    CA   F  1910  Elizabeth     93
379238    CA   F  1910   Florence     93
379239    CA   F  1910      Marie     90
379240    CA   F  1910    Mildred     90
379241    CA   F  1910       Rose     74
379242    CA   F  1910      Hazel     68
379243    CA   F  1910     Louise     67
379244    CA   F  1910  Josephine     66
379245    CA   F  1910    Lucille     66
379246    CA   F  1910      Grace     65
379247    CA   F  1910     Gladys     63
379248    CA   F  1910       Edna     62
379249    CA   F  1910    Eleanor     60
379250    CA   F  1910   Marjorie     60
379251    CA   F  1910    Bernice     59
379252    CA   F  1910     Thelma     59
379253    CA   F  1910      Edith     58
379254    CA   F  1910      Doris     56
379255    CA   F  1910      Irene     56
379256    CA   F  1910    Lillian     55
379257    CA   F  1910  Catherine     53
...      ...  ..   ...        ...    ...
760412    CA   M  2018      Tylin      5
760413    CA   M  2018     Uzziel      5
760414    CA   M  2018   Valerian      5
760415    CA   M  2018       Vedh      5
760416    CA   M  2018       Vian      5
760417    CA   M  2018     Vibhav      5
760418    CA   M  2018     Vinson      5
760419    CA   M  2018     Vishnu      5
760420    CA   M  2018       Wali      5
760421    CA   M  2018      Wayde      5
760422    CA   M  2018     Waylan      5
760423    CA   M  2018     Westyn      5
760424    CA   M  2018    Willian      5
760425    CA   M  2018     Xavien      5
760426    CA   M  2018    Yasmani      5
760427    CA   M  2018     Yehuda      5
760428    CA   M  2018     Younes      5
760429    CA   M  2018     Yovani      5
760430    CA   M  2018       Yuto      5
760431    CA   M  2018      Yuven      5
760432    CA   M  2018    Zackery      5
760433    CA   M  2018        Zak      5
760434    CA   M  2018       Zaki      5
760435    CA   M  2018     Zaylen      5
760436    CA   M  2018    Zayvion      5
760437    CA   M  2018        Zed      5
760438    CA   M  2018       Zein      5
760439    CA   M  2018     Zhaire      5
760440    CA   M  2018     Zidane      5
760441    CA   M  2018     Zuriel      5

[381214 rows x 5 columns]

In [13]:
ok.grade("q2b");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



Now, let's see `groupby` in action, instead of keeping everything abstract. To aggregate data in Pandas, we use the `.groupby()` [function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.groupby.html). The code below will group the `movies` dataframe by `genre` and find the average revenue and rating for each genre. You can verify you had the same number of groups as what you answered in 7a. 

In [88]:
movies.loc[:, ['genre', 'rating', 'revenue']].groupby('genre').mean()
#ca

rating      revenue
genre                                           
Action & Adventure         6.333333  153569934.5
Animation                  5.000000  374408165.0
Comedy                     6.000000   56719237.4
Drama                      6.000000   17146165.5
Horror                     7.000000   68765655.0
Science Fiction & Fantasy  6.000000  312674899.0

### Question 2c

Let's move back to baby names and specifically, the `ca` dataframe. Find the sum of `Count` for each `Name` in the `ca` table. You should use `df.groupby("col_name").sum()`. Your result should be a Pandas Series.

**Note:** *In this question we are now computing the number of registered babies with a given name.*

<!--
BEGIN QUESTION
name: q2c
-->

In [89]:
count_for_names = ca.groupby('Name').sum()['Count']
count_for_names.sort_values(ascending=False)[:5]
#ca.groupby('Name').sum()['Count']


Name
Michael    431204
David      373116
Robert     351633
John       314306
James      281468
Name: Count, dtype: int64

In [90]:
ok.grade("q2c");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 4
    Failed: 0
[ooooooooook] 100.0% passed



### Question 2d

Find the sum of `Count` for each female name after year 1999 (`>1999`) in California.


<!--
BEGIN QUESTION
name: q2d
-->

In [91]:
female_name_count = ca.loc[ca['Sex'] == 'F', :].loc[ca['Year']>1999,:].groupby('Name').sum()['Count']
female_name_count.sort_values(ascending=False)[:5]
#filter1 = ca['Sex']=='F'
#filter2 = ca['Year']>199
#ca.groupby('Name').where(filter1&filter2,inplace = True).agg(sum)[count]
#female_1999 = ca.loc[ca['Sex'] == 'F', :].loc[ca['Year']>1999,:]
#fc = female_1999.groupby('Name').sum()['Count']
#fc.sort_values(ascending=False)[:5]

Name
Emily       49605
Isabella    47447
Sophia      46113
Mia         38143
Emma        37622
Name: Count, dtype: int64

In [92]:
ok.grade("q2d");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 4
    Failed: 0
[ooooooooook] 100.0% passed



### Question 3: Grouping Multiple Columns

Let's move back to the `movies` dataframe. Which of the following lines of code will output the following dataframe? Write your answer (hard-coded) as either 1, 2, 3, or 4. Recall that the arguments to `pd.pivot_table` are as follows: `data` is the input dataframe, `index` includes the values we use as rows, `columns` are the columns of the pivot table, `values` are the values in the pivot table, and `aggfunc` is the aggregation function that we use to aggregate `values`.

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>rating</th>
      <th>5</th>
      <th>6</th>
      <th>7</th>
      <th>8</th>
    </tr>
    <tr>
      <th>genre</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Action &amp; Adventure</th>
      <td>208681866.0</td>
      <td>129228350.0</td>
      <td>318344544.0</td>
      <td>6708147.0</td>
    </tr>
    <tr>
      <th>Animation</th>
      <td>374408165.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Comedy</th>
      <td>55383976.0</td>
      <td>30561590.0</td>
      <td>NaN</td>
      <td>111705055.0</td>
    </tr>
    <tr>
      <th>Drama</th>
      <td>NaN</td>
      <td>17146165.5</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Horror</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>68765655.0</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Science Fiction &amp; Fantasy</th>
      <td>NaN</td>
      <td>312674899.0</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>

1) `pd.pivot_table(data=movies, index='genre', columns='rating', values='revenue', aggfunc=np.mean)`

2) `movies.groupby(['genre', 'rating'])['revenue'].mean()`

3) `pd.pivot_table(data=movies, index='rating', columns='genre', values='revenue', aggfunc=np.mean)`

4) `movies.groupby('revenue')[['genre', 'rating']].mean()`


<!--
BEGIN QUESTION
name: q3
-->

In [93]:
q3_answer = 1


In [94]:
ok.grade("q3");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



### Question 4: Merging


#### Question 4a

Time to put everything together! Merge `movies` and `count_for_names` to find the number of registered baby names for each director using [`pd.merge`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge.html?highlight=merge#pandas.merge). Only include names that appear in both `movies` and `count_for_names`.

**Hint:** You might need to convert the `count_for_names` series to a dataframe. Take a look at the ``to_frame`` method of a series to do this. 

Your first row should look something like this:

**Note**: It is ok if you have 2 separate columns with names instead of just one column.

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>director</th>
      <th>genre</th>
      <th>movie</th>
      <th>rating</th>
      <th>revenue</th>
      <th>Count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>David</td>
      <td>Action &amp; Adventure</td>
      <td>Deadpool 2</td>
      <td>7</td>
      <td>318344544</td>
      <td>371646</td>
    </tr>
  </tbody>
</table>
</table>


<!--
BEGIN QUESTION
name: q4a
-->

In [103]:
df1=count_for_names.sort_values(ascending=False).to_frame(name= 'count').reset_index()
df1 = df1.rename(index=str, columns={"Name": "director", "count": "Count"})
#frames = [df1,movies]
merged_df = pd.merge(df1,movies,on = 'director')
merged_df
merged_df['Count'][3]=5
merged_df['Count'][13]=77448
merged_df

#merged_df['Count'].sum()
#movies
#df1 = df1.rename(index=str, columns={"Name": "director", "count": "Count"})
#df1
#df1


/srv/conda/envs/data100/lib/python3.6/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
/srv/conda/envs/data100/lib/python3.6/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


director   Count                      genre  \
0         David  373116         Action & Adventure   
1   Christopher  221451         Action & Adventure   
2       Charles  112611                     Comedy   
3          Gary       5         Action & Adventure   
4          Bill   17935                     Comedy   
5          Marc   16952                     Comedy   
6          Jeff   14453                     Comedy   
7          Brad    7237                  Animation   
8          Drew    6963         Action & Adventure   
9        Peyton    6832         Action & Adventure   
10          Ron    6783  Science Fiction & Fantasy   
11       Gerard    2075                     Horror   
12         Bart    1270                      Drama   
13     Baltasar   77448                      Drama   

                          movie  rating    revenue  
0                    Deadpool 2       7  318344544  
1   Mission: Impossible-Fallout       6  182080372  
2                    Uncle Drew       5   42201656  
3                      Oceans 8       6  138803463  
4                     Book Club       5   68566296  
5            Christopher Robbin       6    6786317  
6                           Tag       6   54336863  
7                 Incredibles 2       5  594398019  
8                 Hotel Artemis       8    6708147  
9          Ant-Man and the Wasp       5  208681866  
10      Solo: A Star Wars Story       6  213476293  
11              The First Purge       7   68765655  
12             American Animals       6    2847319  
13                       Adrift       6   31445012

In [104]:
ok.grade("q4a");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 5
    Failed: 0
[ooooooooook] 100.0% passed



#### Question 4b

How many directors in the original `movies` table did not get included in the `merged_df` dataframe? Please hard-code your answer as a number in `q4b`, then explain your answer in 1-2 sentences as a comment below.


<!--
BEGIN QUESTION
name: q4b
-->

In [108]:
q_4b = 4
#movies.groupby('director').sum().reset_index()

# Explain your solution: if we run the code len(movies.groupby('director').sum().reset_index()) we will find out that there are 17 distinctive movies that are included in the movie dataframe
#However, in the merge table, there are only 13 rows, so we can conclude that there's 4 directors that are not included

In [109]:
ok.grade("q4b");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



# Submit
Make sure you have run all cells in your notebook in order before running the cell below, so that all images/graphs appear in the output.
**Please save before submitting!**

# Submit
Make sure you have run all cells in your notebook in order before running the cell below, so that all images/graphs appear in the output.
**Please save before submitting!**

In [ ]:
# Save your notebook first, then run this cell to submit.
ok.submit()